---
title: "Report Analysis 16S"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE,
                      warning = FALSE,
                      comment = "",
                      eval=  TRUE,
                      include = TRUE)

```

```{r }
library(phyloseq)
library(vegan)
library(pairwiseAdonis)
library(microbiome)
library(microbiomeMarker)
library(qiime2R)
library(ggplot2)
library(ggpubr)
library(ggVennDiagram)
library(kableExtra)
library(pander)
library(dplyr)
library(magrittr)
library(VennDiagram)
library(multcompView)
library(forcats)
library(DESeq2)
```

```{r echo=FALSE}
#load own functions
source("../M.senegalensis-core-microbiota/functions.R")
```

# Experimental design

The objective of this study is to analyze whether soil microbial communities co-evolve with *Maytenus senegalensis*. To this end,we have compared two scenarios: soils under the canopy of *Maytenus senegalensis* shrubs (Understory condition) and soils without vegetation (Gap condition), both in Senegal and Spain.

We have performed an amplicon 16S an analysis to investigate microbial communities composition in each soil type. Moreover, we aimed to identify if there are certain microorganisms that are associated to Understory condition in both Senegal and Spain. With this approach we aim to use amplicon technology to look for those microorganisms that co-evolve with *Maytenus senegalensis*.


# Load Data
  
The sequences were analyse with QIIME2 software. We are going to load the different datasets with the help of qiime2R package and build a phyloseq object. To this end we need:  
- Feature table  
- Taxonomy table  
- Rooted tree file  
- Metadata file  

```{r }
physeq <-qza_to_phyloseq(features="taxonomy/final_FeatureTable.qza", 
                         tree="phylogeny/rooted-tree.qza", 
                         taxonomy="taxonomy/taxonomy.qza", 
                         metadata="metadata.txt")
```

# Negative control analysis

To detect possibles contamination in the samples we have included a negative control test made up by distilled water.
  
```{r controltaxonomy}
sample_data(physeq)$SampleID <- row.names(sample_data(physeq))
sample_data(physeq)$SampleType <- paste0(sample_data(physeq)$Country,"_" , sample_data(physeq)$Microhabitat)
#obtain negative control table
neg_ps <- subset_samples(physeq, SampleID=="KitControl40")
#remove ASVs == 0
neg_ps.clean = prune_taxa(taxa_sums(neg_ps) > 1, neg_ps)

kable(tax_table(neg_ps.clean), row.names = FALSE, caption = "Negative control taxonomy.")
```

The table shows an ASV in the negative control.


```{r contaminationanalysis}
#Obtain the ASVs within the negative control
contan_asv <- rownames(otu_table(neg_ps.clean))
#Count the number of samples in which the otu is present
if(length(contan_asv) > 0){
  contan_asv_sum <- rowSums(otu_table(physeq)[contan_asv] != 0)
  kable(contan_asv_sum, col.names = "Sample", caption = "Number of samples with the contaminant ASVs.")
}else{
  print("There are not any contaminant ASV in the sample.")
}
```

The bacteria is only available in negative control sample, so this will be removed in the quality control filter.

# Positive control - Mock Analysis 

As a quality control measure, a positive control with known microorganisms was introduced. Specifically, it contains eight bacteria:  
- ***Bacillus subtilis*** (G+)  
- ***Listeria monocytogenes*** (G+)  
- ***Staphylococcus aureus*** (G+)  
- ***Enterococcus faecalis*** (G+)  
- ***Lactobacillus fermentum*** (G+)  
- ***Salmonella enterica*** (G-)  
- ***Escherichia coli*** (G-)   
- ***Pseudomonas aeruginosa*** (G-) 

Therefore, the first thing we are going to check is that we have been able to taxonomically identify these microorganisms in the mock sample.

```{r mocktaxonomy}
#obtain mock table
mock_ps <- subset_samples(physeq, SampleID=="MockZymo41")
#remove ASVs == 0
mock_ps.clean = prune_taxa(taxa_sums(mock_ps) > 1, mock_ps)
#agglomerate taxonomic results to Genus level
mock_tg <- tax_glom(mock_ps.clean, taxrank="Genus")
mock_tf <- tax_glom(mock_ps.clean, taxrank="Family")

#obtain taxonomic table
kable(tax_table(mock_tg)[,1:6], row.names = FALSE, caption = "Mock taxonomy at Genus level.")
```

This table shows that we were able to perform a taxonomic assignment for seven bacteria at genus level. However, we couldn't find *Salmonella enterica* at genus level but it is also found at family level as it belongs to Enterobacteriaceae family.


# Quality Analysis
  
Before conducting downstream analysis, we need to perform some quality analysis and apply some filters in the case it is necessary. In this analysis we are keeping Archea and Bacteria data.

## Eliminate Negative and Positive control ASVs  
  
The first step in our quality control analysis is to remove the ASVs from our negative and positive controls and the singletons.

```{r RemoveControl}
#remove mock sample
physeq_wo_mock <- subset_samples(physeq, !SampleID %in% c("MockZymo41", "KitControl40"))
#remove ASVs == 1
physeq.clean <- prune_taxa(taxa_sums(physeq_wo_mock) > 1, physeq_wo_mock)
```

## Sequencing Depth

Create histograms that summarize sequencing depth in our data. Start with overall, but then also create Microhabitat specific plots to help evaluate the balance in the experiment. If there are samples with very reads will be considered to be removed.

```{r ReadsPerSample, fig.cap="Distribution of reads per sample."}
#obtain number of read per sample
TotalReads <- sample_sums(physeq.clean)
readcount <- cbind(sample_data(physeq.clean), TotalReads)
#head(readcount[order(readcount$TotalReads), c("SampleID", "TotalReads")])

#Plot the distribution of the counts of our samples
ggplot(readcount, aes(TotalReads)) + geom_histogram() + ggtitle("Sequencing Depth") +
           theme(panel.border = element_rect(colour = "black", fill=NA, size=1), axis.title=element_text(size=14,face="bold"), axis.text=element_text(size=12), legend.text=element_text(size=12))
```  


## Rarefaction curves

Rarefaction curve could be used to determined whether the sequencing depth cover microbial diversity of the samples.
  
```{r rarecurve, fig.cap="Rarefaction curves based on sequencing depth. The graph shows rarefaction curves for each sample, plotting the number of observed species (y-axis) as a function of the number of sequencing reads which represent sample sizes (x-axis)."}
otu_t <- otu_table(physeq.clean)
otu_df <- as.data.frame(t(otu_t))
sample_names <- rownames(otu_df)
# we will use vegan rarecurve 
otu.rarecurve <- rarecurve(otu_df, step = 10000, label = F)
```  

We can see that the size of the libraries is very different in each samples. However, we are not going to perform a rarefaction, as we want to analyse the core microbiota in further steps, so we would lose data. However, we need to adjust the data so that we can compare different samples without being influenced by the total number of readings taken.
  
## Normalization

Generally, sequencing data sets are not of the same size. So, to make relatively fair comparisons between sets, we need to normalize the counts. It is necessary to prepare the data in a way that improves the comparability of the samples. There are different methods to perform normalization, but in this case we are using Cumultive Sum Scaling with *microbiomeMarker* package.

```{r PhyloseqNormalizarion}
#cumulative sum scaling
physeq.css <- normalize(physeq.clean, "CSS")
```

## Filtering

ASVs with very few readings may be technical noise (sequencing errors, contamination, or the presence of extremely rare taxa that are not relevant to the overall analysis). We filter out ASVs present in less than 10% of the samples.  

```{r DataFiltering}
#remove ASVs not in 10% of the samples
physeq.10 <- prune_taxa(rowSums(otu_table(physeq.css) != 0) > ncol(otu_table(physeq.css)) * 0.1, physeq.css)
```


# Count the number of Taxa per experimental conditions

In this first analysis, we count the number of taxa in the different microhabitats at different taxonomic levels and test whether they are significantly different. Here we are not taking into account the abundance of the different taxa.

**At Genus level:**

```{r ObtainGenusData}
#agglomerate taxonomic results to Genus level
ps.genus_all <- tax_glom(physeq.10, taxrank="Genus")
genus_l <- get_taxa_unique(ps.genus_all, taxonomic.rank="Genus")

#remove unasigned genera
genus_to_remove <- c(NA, "uncultured")
genus_l <- genus_l[!genus_l %in% genus_to_remove]
ps.genus <- subset_taxa(ps.genus_all, Genus %in% genus_l)

#Obtain a df and remove abundance 0
genus_df <- psmelt(ps.genus)
genus_df <- genus_df[!genus_df$Abundance == 0,]

#Count the number of unique Genera per sample
genus_count_sample_df <- genus_df %>%
  group_by(SampleID) %>%
  summarise(Genus_nb = length(unique(Genus)))
genus_count_sample_df$Country <- sample_data(physeq.clean)$Country
genus_count_sample_df$Microhabitat <- sample_data(physeq.clean)$Microhabitat

#ANOVA
anova <- aov(Genus_nb ~ Country * Microhabitat, data = genus_count_sample_df)
pander(summary(anova))

#Post-hoc Tukey
tukey <- TukeyHSD(anova)
letters <- multcompLetters(tukey$`Country:Microhabitat`[, "p adj"])$Letters
letters <- as.data.frame(letters)
new_cols <- do.call(rbind, strsplit(rownames(letters), ":")) 
letters$Country <- new_cols[,1]
letters$Microhabitat <- new_cols[,2]

#Mean and SD of the number of Genera per Country and Microhabitat
genus_count_df <- genus_count_sample_df %>%
  group_by(Country, Microhabitat) %>%
  summarise(Genus_mean = mean(Genus_nb), SD = sd(Genus_nb))

genus_count_df <- merge(genus_count_df, letters, by.x = c("Country", "Microhabitat"), by.y = c("Country", "Microhabitat"))

kable(genus_count_df, row.names = FALSE, caption = "Number of genera identified per experimental condition. The table presents the total number of genera detected in each experimental condition (Understory vs Gap), without considering their abundances. This provides an overview of the taxonomic diversity present in each condition based solely on the presence of genera.")
```
  
As we can see in the table there are not significant differences in terms of the number of different genera in the microhabitats of the same country, but there are between the different countries.
  

**At Phylum level**

```{r ObtainPhylumData}
#agglomerate taxonomic results to Phylum level
ps.phylum <- tax_glom(physeq.10, taxrank="Phylum")

#Obtain a df and remove abundance 0
phylum_df <- psmelt(ps.phylum)
phylum_df <- phylum_df[!phylum_df$Abundance == 0,]

#Count the number of unique phyla per sample
phylum_count_sample_df <- phylum_df %>%
  group_by(SampleID) %>%
  summarise(Phylum_nb = length(unique(Phylum)))
phylum_count_sample_df$Country <- physeq.clean@sam_data$Country
phylum_count_sample_df$Microhabitat <- physeq.clean@sam_data$Microhabitat

#ANOVA
anova <- aov(Phylum_nb ~ Country * Microhabitat, data = phylum_count_sample_df)
pander(summary(anova))

#Post-hoc Tukey
tukey <- TukeyHSD(anova)
letters <- multcompLetters(tukey$`Country:Microhabitat`[, "p adj"])$Letters
letters <- as.data.frame(letters)
new_cols <- do.call(rbind, strsplit(rownames(letters), ":")) 
letters$Country <- new_cols[,1]
letters$Microhabitat <- new_cols[,2]

#Mean and SD of the number of Families per Country and Microhabitat
phylum_count_df <- phylum_count_sample_df %>%
  group_by(Country, Microhabitat) %>%
  summarise(Phylum_mean = mean(Phylum_nb), SD = sd(Phylum_nb))

phylum_count_df <- merge(phylum_count_df, letters, by.x = c("Country", "Microhabitat"), by.y = c("Country", "Microhabitat"))

kable(phylum_count_df, row.names = FALSE, caption = "Number of Phylum identified per experimental condition. The table presents the total number of Phylum detected in each experimental condition (Understory vs Gap), without considering their abundances. This provides an overview of the taxonomic diversity present in each condition based solely on the presence of phyla")
```
  
As we can see in the table there are not significant differences in terms of the number of different genera in the microhabitats of the same country, but there are between the different countries.

# Alpha Diversity
  
Alpha diversity refers to the number of species or the diversity of a community in a single sample. It can be measured using various indices:  

-**Chao**: Measures species richness, including both observed and undetected species, by estimating the number of species likely present in a community based on the frequency of rare species.  


```{r prepare data}
#STABLISH WHICH ALPHA DIVERSITY METRICS DO YOU WANT TO CALCULATE
metrics <- c("Chao1")

#STABLIS THE COMPARISONS BETWEEN GROUPS DO YOU WANT TO CALCULATE
my_comparisons_microhabitat <-list(c("Understory", "Gap"))
my_comparisons_country <-list(c("Senegal", "Spain"))
#STABLISH THE STATISTIVAL METRICS TO CALCULATE IF THE DIVERSITY IS SIFNIFICANT BETWEEN GROUPS
stat_metrics <- c("kruskal.test")

```
  
## Senegal Understory vs Senegal Gap

```{r AlphaDiversitySenegal, fig.width=8, fig.height=6}
#Senegal data
physeq.senegal <- subset_samples(physeq.10, Country == "Senegal")
physeq.senegal <- prune_taxa(taxa_sums(physeq.senegal) > 1, physeq.senegal)

#Chao Plot
alpha_bact_senegal <- alpha_diversity_plot(physeq.senegal, "Microhabitat", metrics, my_comparisons_microhabitat, stat_metrics, asterisk = FALSE, "Senegal") + theme(strip.text = element_blank(), axis.title.y=element_blank(),  legend.position = "none", axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
alpha_bact_senegal

#alpha diversity table
senegal.tb <- estimate_richness(physeq.senegal, measures=metrics)
senegal.tb$Microhabitat <- sample_data(physeq.senegal)$Microhabitat
senegal.tb$Country <- rep("Senegal", nrow(sample_data(physeq.senegal)))
kable(senegal.tb)
```

The alpha diversity between both experimental conditions is not significantly different with the Kruskal wallys test and a p-value adjustment with BH. This indicates that within each group or condition, the communities have not different diversity in terms of the number of species or relative abundance. In other words, there are not differences in the quantity or abundance of species within each condition.  

## Spain Understory vs Spain Gap  
  
```{r AlphaDiversitySpain, fig.width=8, fig.height=6}
#Spain data
physeq.spain <- subset_samples(physeq.10, Country == "Spain")
physeq.spain <- prune_taxa(taxa_sums(physeq.spain) > 1, physeq.spain)

#Chao Plot
alpha_bact_spain <- alpha_diversity_plot(physeq.spain, "Microhabitat", metrics, my_comparisons_microhabitat, stat_metrics, asterisk = FALSE, "Spain") + theme(strip.text = element_blank(), axis.title.y=element_blank(), legend.position = "none", axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) 
alpha_bact_spain

#alpha diversity table
spain.tb <- estimate_richness(physeq.spain, measures=metrics)
spain.tb$Microhabitat <- sample_data(physeq.spain)$Microhabitat
spain.tb$Country <- rep("Spain", nrow(sample_data(physeq.spain)))
kable(spain.tb)
```

The alpha diversity between both experimental conditions is not significantly different with the Kruskal wallys test and a p-value adjustment with BH. This indicates that within each group or condition, the communities have not different diversity in terms of the number of species or relative abundance. In other words, there are not differences in the quantity or abundance of species within each condition.    
  
## Senegal Understory vs Spain Understory   
  
```{r AlphaDiversityUnderstory, fig.width=8, fig.height=6}
#Understory data
physeq.understory <- subset_samples(physeq.10, Microhabitat == "Understory")
physeq.understory <- prune_taxa(taxa_sums(physeq.understory) > 1, physeq.understory)

#Chao Plot
alpha_bact_under <- alpha_diversity_plot(physeq.understory, "Country", metrics, my_comparisons_country, stat_metrics, asterisk = FALSE, "Under") + theme(strip.text = element_blank(), axis.title.y=element_blank(), legend.position = "none", axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) + scale_fill_manual(values = c("Spain" = "#E69F00", "Senegal" = "#C3D7A4"))
alpha_bact_under

#alpha diversity table
understory.tb <- estimate_richness(physeq.understory, measures=metrics)
understory.tb$Country <- sample_data(physeq.understory)$Country
understory.tb$Microhabitat <- rep("Understory", nrow(sample_data(physeq.understory)))
kable(understory.tb)
```

The alpha diversity between both experimental conditions is significantly different in the three indices with the Kruskal wallys test and a p-value adjustment with BH. This indicates that within each group or condition, the communities have different diversity in terms of the number of species or relative abundance. In other words, there are some differences in the quantity or abundance of species within each condition.

## Senegal Gap vs Spain Gap   

```{r AlphaDiversityGap, fig.width=8, fig.height=6}
#Gap data
physeq.gap <- subset_samples(physeq.10, Microhabitat == "Gap")
physeq.gap <- prune_taxa(taxa_sums(physeq.gap) > 1, physeq.gap)

#Chao Plot
alpha_bact_gap <- alpha_diversity_plot(physeq.gap, "Country", metrics, my_comparisons_country, stat_metrics, asterisk = FALSE, "Gap") + theme(strip.text = element_blank(), axis.title.y=element_blank(), legend.position = "none", axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) + scale_fill_manual(values = c("Spain" = "#E69F00", "Senegal" = "#C3D7A4"))
alpha_bact_gap

#alpha diversity table
gap.tb <- estimate_richness(physeq.gap, measures=metrics)
gap.tb$Country <- sample_data(physeq.gap)$Country
gap.tb$Microhabitat <- rep("Gap", nrow(sample_data(physeq.gap)))
kable(gap.tb)
```

The alpha diversity between both experimental conditions is significantly different for Chao and Shannon Index, but not with Simpson with the Kruskal wallys test and a p-value adjustment with BH. This indicates that within each group or condition, the communities may have different diversity in terms of the number of species or relative abundance but the relevance is based on rare or low abundant taxa. 

# Beta Diversity

Beta diversity measures the differences in species composition between communities or samples. In other words, it assesses how similar or different communities are from each other in terms of their biological composition. Just like with alpha diversity, there are also various metrics that can be used to measure beta diversity. In this case we used **Bray Curtis** dissmilarity metric, which measures differences in species composition and takes into account the relative abundance of species. The more different the two communities, the greater the distance.   

To visualize beta diversity we used a Principal Coordinates Analysis (**PCoA**). Axis 1 (the first principal component) captures the maximum possible variation in the data between samples. The further apart two points are on Axis 1, the more different they are in terms of species composition, based on the dissimilarity or distance metric used. Axis 2 (the second principal component) represents the second greatest amount of variation between samples that was not captured by Axis 1. 

This method only provide a graphical representation of the differences between samples and help identify patterns or groupings in the data. However, it is necessary to assess whether the differences in species composition between groups (or conditions) are statistically significant. **PERMANOVA** evaluates whether the differences in species composition between groups are statistically significant, based on the variability of dissimilarities between and within groups. The results include a p-value that indicates whether the differences between groups in terms of species composition are significant. We have include pairwise PERMANOVA and a permutest as a Post Hoc analysis.  

***Bray-curtis dissimilarity***

```{r BraycurtisDistanceCalculation}
distance <- "bray"

set.seed(1)

bray_distance <- phyloseq::distance(physeq.10, method = distance)
meta <- data.frame(sample_data(physeq.10))
```

**PERMANOVA**


```{r BraycurtisPERMANOVAAllSamples}
bray_permanova <- adonis2(bray_distance ~ SampleType, data = meta, permutations = 999, method = distance)

#create label for the plots
bp_pvalue <- paste("P-value", bray_permanova$`Pr(>F)`[1])

capture.output(bray_permanova, file = "permanova_16S_braydistance.txt")
bray_permanova_text <- paste("PERMANOVA", bp_pvalue, sep = "\n")
pander(bray_permanova)
```

## Post Hoc

### Pairwise PERMANOVA

**Senegal**  
  
```{r BraycurtisPERMANOVASenegal}
pairwise_bray <- pairwise.adonis2(bray_distance ~ SampleType, data = meta)
pander(pairwise_bray$`Senegal_Understory_vs_Senegal_Gap`)
```

**Spain**  
  
```{r BraycurtisPERMANOVASpain}
pander(pairwise_bray$`Spain_Understory_vs_Spain_Gap`)
```

**Understory**  
  
```{r BraycurtisPERMANOVAUnderstory}
pander(pairwise_bray$`Spain_Understory_vs_Senegal_Understory`)
```

**Gap**  
  
```{r BraycurtisPERMANOVAGap}
pander(pairwise_bray$`Spain_Gap_vs_Senegal_Gap`)
```

We performed an analysis to test the homogeneity condition.  

```{r BraycurtisPERMANOVPostHoc}
#Checking the homogeneity condition
homogeneity_bray_Country <- permutest(betadisper(bray_distance, c(meta$Country)), pairwise = TRUE)
pander(homogeneity_bray_Country$tab)

homogeneity_bray_Microhabitat <- permutest(betadisper(bray_distance, c(meta$Microhabitat)), pairwise = TRUE)
pander(homogeneity_bray_Microhabitat$tab)
```
  
A **significant PERMANOVA** indicates that there are significant differences in species composition between groups, based on the variability in dissimilarity between the groups. A **non-significant permutest** may suggest that, although the compositions of the groups are significantly different, the dispersion or variability within the groups is not significantly different between them. This may occur if the differences in composition between the groups are large enough to be detected by PERMANOVA, but the dispersion within each group does not significantly contribute to these differences. It is possible that the groups differ in terms of species composition (PERMANOVA), but the internal variability within each group is similar (non-significant permutest). In other words, each group may have a distinct species composition, but the variability in that composition within each group is comparable.  
  
***PCoA - Bray Curtis*** 

```{r BraycurtisPCoA, fig.cap="Beta diversity analysis using Bray-Curtis dissimilarity visualized by PCoA. The plot represents the principal coordinates analysis (PCoA) based on Bray-Curtis distances, illustrating the dissimilarity between samples across experimental groups. A PERMANOVA test was conducted to assess statistical differences, with the p-value displayed on the graph."}
method <- "PCoA"

pcoa_bray <- betadiversity_analysis(physeq.10, method, distance)
beta_bact <- plot_distance(physeq.10, pcoa_bray, "Microhabitat", shape="Country", elipse = TRUE) + annotate("text" ,x=0.2,y=0.3, label = bray_permanova_text, size = 5) 
beta_bact
```

# Composition Analysis

A composition plot of an amplicon typically refers to a graphical representation used to visualize the abundance or distribution of the microorganisms community at a taxonomic level.  

## Phylum Pie Chart

```{r PhylumAbundancePieChart, fig.width=8, fig.cap="Phylum Relative Abundance"}
#Caluclate Relative Abundance
ps.phylum.rel <- transform_sample_counts(ps.phylum, function(x){x / sum(x)})
df.phylum.rel <- psmelt(ps.phylum.rel)

phylum_sumarize <- df.phylum.rel %>%  group_by(Country, Microhabitat, Phylum) %>% 
  summarise(Abundance_mean=mean(Abundance, na.rm=TRUE), sd=sd(Abundance, na.rm=TRUE)) 
phylum_sumarize <- phylum_sumarize[!phylum_sumarize$Abundance_mean == 0,]

phy_piechart <- plot_piechart(phylum_sumarize, "Abundance_mean", "Phylum") +  facet_wrap(Country~Microhabitat, scales = "free") 
phy_piechart
```

Despite there are the same phyla in all the microhabitats, the proportion of them are slightly different.

## Top 20 Genus

```{r TopGenus, fig.width=15, fig.height=8}
ps.genus.rel <- transform_sample_counts(ps.genus, function(x){x / sum(x)})
ps.genus.rel@phy_tree <- NULL
#Obtain top 20 most abundant genera
top.20 <- top_taxa(ps.genus.rel, n = 20)
ps.top <- prune_taxa(top.20, ps.genus.rel)
#Create a dataframe from phyloseq object
df_top <- psmelt(ps.top)
names(df_top)[3] <- paste("Rel.Abundance")

#Calculate the mean and sd of the abundance per Genus, Country and experimental condition
top_sumarize <- df_top %>%  group_by(Country, Microhabitat, Genus) %>% 
  summarise(Relative_Abundance=mean(Rel.Abundance), sd=sd(Rel.Abundance), se = sd(Rel.Abundance) / sqrt(n()), .groups = "drop")

top_bact <- ggplot(top_sumarize, aes(x = Relative_Abundance, y = Genus, fill = Microhabitat)) +
  geom_bar(stat = "identity", position = position_dodge(width = 0.8), width = 0.7) +
  geom_errorbar(aes(xmin=Relative_Abundance, xmax=Relative_Abundance+se),
                  width=.5,                    # Width of the error bars
                  position=position_dodge(.9)) +
  facet_wrap(~ Country, scales = "free") +
  theme_minimal() +
  theme(
  axis.text=element_text(size=20), 
  axis.title=element_text(size=20,face="bold"),
  legend.text=element_text(size=20),
  legend.title = element_text(size=20),
  strip.text = element_text(size = 22, face = "bold")
  )
top_bact
```

## Differencial Abundance  between microhabitats

```{r DAASenegal}
physeq.senegal.genus <- tax_glom(physeq.senegal, taxrank="Genus")
senegal_genus_l <- get_taxa_unique(physeq.senegal.genus, taxonomic.rank = "Genus")
senegal_genus_l <- senegal_genus_l[!senegal_genus_l %in% genus_to_remove]
physeq.senegal.genus <- subset_taxa(physeq.senegal.genus, Genus %in% senegal_genus_l)
alpha <- 0.05
ds_senegal <- phyloseq_to_deseq2(physeq.senegal.genus, ~ Microhabitat)


diagdds_senegal <-  DESeq(ds_senegal, 
                  test="Wald", 
                  fitType="parametric", 
                  sfType = 'poscounts')
res.senegal <- results(diagdds_senegal, cooksCutoff = F,
               contrast = c("Microhabitat", "Understory", "Gap") )
sigtab.senegal <- res.senegal[which(res.senegal$padj < alpha), ]
sigtab.senegal <- cbind(as(sigtab.senegal, "data.frame"), as(tax_table(physeq.senegal.genus)[rownames(sigtab.senegal), ], "matrix"))

senegal.daa.bact.barplot <- daa_bar_plot(sigtab.senegal, contrast, "Senegal      \nGap // Understory", col_pal = FALSE) + 
  theme(
  plot.title = element_text(hjust = 0.9, size = 18), 
  axis.text=element_text(size=12), 
  axis.title=element_text(size=14,face="bold"),
  legend.text=element_text(size=12)
  )
```

```{r DAASpain}
physeq.spain.genus <- tax_glom(physeq.spain, taxrank="Genus")
spain_genus_l <- get_taxa_unique(physeq.spain.genus, taxonomic.rank = "Genus")
spain_genus_l <- spain_genus_l[!spain_genus_l %in% genus_to_remove]
physeq.spain.genus <- subset_taxa(physeq.spain.genus, Genus %in% spain_genus_l)
alpha <- 0.05
ds_spain <- phyloseq_to_deseq2(physeq.spain.genus, ~ Microhabitat)


diagdds_spain <-  DESeq(ds_spain, 
                  test="Wald", 
                  fitType="parametric", 
                  sfType = 'poscounts')
res.spain <- results(diagdds_spain, cooksCutoff = F,
               contrast = c("Microhabitat", "Understory", "Gap") )
sigtab.spain <- res.spain[which(res.spain$padj < alpha), ]
sigtab.spain <- cbind(as(sigtab.spain, "data.frame"), as(tax_table(physeq.spain.genus)[rownames(sigtab.spain), ], "matrix"))

spain.daa.bact.barplot <- daa_bar_plot(sigtab.spain, contrast, "Spain       \nGap // Understory", col_pal = FALSE) + 
  theme(
  plot.title = element_text(hjust = 0.9, size = 18), 
  axis.text=element_text(size=12), 
  axis.title=element_text(size=14,face="bold"),
  legend.text=element_text(size=12)
  )
```

```{r DAABarplotCountry, fig.width=12}
ggarrange(senegal.daa.bact.barplot, spain.daa.bact.barplot, widths = c(1, 0.8))
```

# Analysis Core Microbiota

Despite the microorganisms community structure is different, there are some genera that are common to the different conditions. To test the co-evolution hypothesis, we need to obtain some genus that are available in all the samples of Understory microhabitat in both countries, Senegal and Spain. This analysis is performed following this steps:    

To find the common genera exclusive from understory we need to perform several steps:  
**1:** Obtain a subset of the samples in understory and gap condition.  
**2:** Agglomerate ASVs at Genus level.  
**3:** Calculate relative abundance of each genera.  
**4:** Filter low abundant taxa (less than 0.2%).  
**5:** Eliminate from understory all the genera that are available in gap samples.  
**6:** Retain those Genera in understory that are within 100% of the samples.  
**7:** Find common genera between Senegal and Spain Understories.
  
```{r SenegalAbundance}
#Obtain A subset from Senegal Gap Samples
ps.senegal.gap <- subset_samples(physeq.css, Microhabitat == "Gap" & Country == "Senegal")
ps.senegal.gap.clean <- prune_taxa(taxa_sums(ps.senegal.gap) > 1, ps.senegal.gap)
ps.senegal.gap.genus <- tax_glom(ps.senegal.gap.clean, taxrank="Genus")
#Calculate relative abundance
ps.senegal.gap.genus_rel = transform_sample_counts(ps.senegal.gap.genus, function(x){(x / sum(x))*100})
#obtain mean abundance per genus
df.senegal.gap <- psmelt(ps.senegal.gap.genus_rel)
df.senegal.gap_mean <- df.senegal.gap %>% group_by(Genus) %>% summarize(Gap=mean(Abundance, na.rm=TRUE))

#Obtain A subset from Senegal Understory Samples
ps.senegal.understory <- subset_samples(physeq.css, Microhabitat == "Understory" & Country == "Senegal")
ps.senegal.understory.clean <- prune_taxa(taxa_sums(ps.senegal.understory) > 1, ps.senegal.understory)
ps.senegal.understory.genus <- tax_glom(ps.senegal.understory.clean, taxrank="Genus")
#Calculate relative abundance 
ps.senegal.understory.genus_rel = transform_sample_counts(ps.senegal.understory.genus, function(x){(x / sum(x))*100})
#obtain mean abundance per genus
df.senegal.understory <- psmelt(ps.senegal.understory.genus_rel)
df.senegal.understory_mean <- df.senegal.understory %>% group_by(Genus) %>% summarize(Understory=mean(Abundance, na.rm=TRUE))

df.senegal.mean <- merge(df.senegal.gap_mean, df.senegal.understory_mean, by = "Genus") %>%
 replace(is.na(.), 0)

plt_senegal <- ggplot(df.senegal.mean, aes(x = sqrt(Understory), y = sqrt(Gap))) +
  geom_point(alpha = 0.7) +
  geom_abline(slope = 1, intercept = 0, color = "red") + 
  geom_vline(xintercept = sqrt(0.2), linetype = "dashed", color = "darkgreen", size = 1) +
  geom_hline(yintercept = sqrt(0.2), linetype = "dotted", color = "blue", size = 1) +
  labs(x = "Relative Abundance Understory", 
       y = "Relative Abundance Gap", 
       title = "Senegal") +
  theme_classic()

```

```{r SpainAbundance}
#Obtain A subset from Spain Gap Samples
ps.spain.gap <- subset_samples(physeq.css, Microhabitat == "Gap" & Country == "Spain")
ps.spain.gap.clean <- prune_taxa(taxa_sums(ps.spain.gap) > 1, ps.spain.gap)
ps.spain.gap.genus <- tax_glom(ps.spain.gap.clean, taxrank="Genus")
#Calculate relative abundance
ps.spain.gap.genus_rel = transform_sample_counts(ps.spain.gap.genus, function(x){(x / sum(x))*100})
#obtain mean abundance per genus
df.spain.gap <- psmelt(ps.spain.gap.genus_rel)
df.spain.gap_mean <- df.spain.gap %>% group_by(Genus) %>% summarize(Gap=mean(Abundance, na.rm=TRUE))

#Obtain A subset from spain Understory Samples
ps.spain.understory <- subset_samples(physeq.css, Microhabitat == "Understory" & Country == "Spain")
ps.spain.understory.clean <- prune_taxa(taxa_sums(ps.spain.understory) > 1, ps.spain.understory)
ps.spain.understory.genus <- tax_glom(ps.spain.understory.clean, taxrank="Genus")
#Calculate relative abundance and filter out low abundant genera
ps.spain.understory.genus_rel = transform_sample_counts(ps.spain.understory.genus, function(x){(x / sum(x))*100})
#obtain mean abundance per genus
df.spain.understory <- psmelt(ps.spain.understory.genus_rel)
df.spain.understory_mean <- df.spain.understory %>% group_by(Genus) %>% summarize(Understory=mean(Abundance, na.rm=TRUE))

df.spain.mean <- merge(df.spain.gap_mean, df.spain.understory_mean, by = "Genus") %>%
 replace(is.na(.), 0)

plt_spain <- ggplot(df.spain.mean, aes(x = sqrt(Understory), y = sqrt(Gap))) +
  geom_point(alpha = 0.7) +
  geom_abline(slope = 1, intercept = 0, color = "red") + 
  geom_vline(xintercept = sqrt(0.2), linetype = "dashed", color = "darkgreen", size = 1) +
  geom_hline(yintercept = sqrt(0.2), linetype = "dotted", color = "blue", size = 1) +
  labs(x = "Relative Abundance Understory", 
       y = "Relative Abundance Gap", 
       title = "Spain") +
  theme_classic()
```

```{r AbundanceComparisonPlot, fig.width = 10, fig.cap="Compare relative abundance between understory and Gap samples. Points above red line means relative abundance is higher in gap. Points bellow red line means the relative abundance is higher in understory. Green lines correspond to abundance filter"}

plt <- ggarrange(plt_senegal + rremove("ylab") + rremove("xlab"), plt_spain+ rremove("ylab") + rremove("xlab"), ncol = 2)

annotate_figure(plt, left = textGrob("Relative Abundance Gap", rot = 90, gp = gpar(fontsize = 12, fontface = "bold")),
                bottom = textGrob("Relative Abundance Understory", gp = gpar(fontsize = 12, fontface = "bold")))
```

  
**Exclusive Senegal**  
  
```{r SenegalVenn}
#Filter Senegal Gap by relative abundance
ps.senegal.gap.genus1 <- filter_taxa(ps.senegal.gap.genus_rel, function(x){ mean(x) >= 0.2}, prune = TRUE)
#Obtain Senegal gap genus list
ps.senegal.gap.l <- get_taxa_unique(ps.senegal.gap.genus1, taxonomic.rank="Genus")

#Filter Senegal Understory by relative abundance
ps.senegal.understory.genus1 <- filter_taxa(ps.senegal.understory.genus_rel, function(x){ mean(x) >= 0.2}, prune = TRUE)
```

```{r SenegalExclusive}
#Remove Gap genera in Understory
senegal_understory_exclusive <- subset_taxa(ps.senegal.understory.genus1, !Genus %in% ps.senegal.gap.l)
senegal_understory_exclusive.l <- get_taxa_unique(senegal_understory_exclusive, taxonomic.rank="Genus")

#Obtain Genera that are in the 100% samples
senegal_understory_exclusive.all <- prune_taxa(rowSums(otu_table(senegal_understory_exclusive) != 0) == ncol(otu_table(senegal_understory_exclusive)), senegal_understory_exclusive)

pander(paste("There are ", length(senegal_understory_exclusive.l)," exclusive genera in Senegal understory: ",paste(senegal_understory_exclusive.l, collapse = ", ")))  
pander(paste("Among all the exclusive genera in Senegal understory,", length(get_taxa_unique(senegal_understory_exclusive.all, taxonomic.rank="Genus"))," are within all the samples: ",paste(get_taxa_unique(senegal_understory_exclusive.all, taxonomic.rank="Genus"), collapse = ", ")))  
```
  
**Exclusive Spain**  
  

```{r SpainVenn}
#Filter Spain Gap by relative abundance
ps.spain.gap.genus1 <- filter_taxa(ps.spain.gap.genus_rel, function(x){ mean(x) >= 0.2}, prune = TRUE)
#Obtain list of Spain Gap genera
ps.spain.gap.l <- get_taxa_unique(ps.spain.gap.genus1, taxonomic.rank="Genus")

#Filter Spain Understory by relative Abundance
ps.spain.understory.genus1 <- filter_taxa(ps.spain.understory.genus_rel, function(x){ mean(x) >= 0.2}, prune = TRUE)

```

```{r SpainExclusive}
#Remove Gap genera
spain_understory_exclusive <- subset_taxa(ps.spain.understory.genus1, !Genus %in% ps.spain.gap.l)
spain_understory_exclusive.l <- get_taxa_unique(spain_understory_exclusive, taxonomic.rank="Genus")
pander(paste("There are ", length(spain_understory_exclusive.l)," exclusive genera in Spain understory: ",paste(spain_understory_exclusive.l, collapse = ", ")))

#Obtain Genera that are in the 100% samples
spain_understory_exclusive.all <- prune_taxa(rowSums(otu_table(spain_understory_exclusive) != 0) == ncol(otu_table(spain_understory_exclusive)), spain_understory_exclusive)
pander(paste("Among all the exclusive genera in Spain understory,", length(get_taxa_unique(spain_understory_exclusive.all, taxonomic.rank="Genus"))," are within all the samples: ",paste(get_taxa_unique(spain_understory_exclusive.all, taxonomic.rank="Genus"), collapse = ", ")))  
```
  
**Core Microbiota**  

We want to compare if there are shared taxa between every microhabitat:  

```{r AllVenn}
lst3 <- list(Sp.Gap= ps.spain.gap.l,
            Sp.Unde= get_taxa_unique(ps.spain.understory.genus1, taxonomic.rank="Genus"),
            Sn.Gap= ps.senegal.gap.l,
            Sn.Unde= get_taxa_unique(ps.senegal.understory.genus1, taxonomic.rank="Genus"))

ven_bact <- ggVennDiagram(lst3, label_alpha = 0, set_size = 6, label_size = 6) +  scale_fill_gradient(low = "#F4FAFE", high = "#4981BF") +  theme(legend.position = "none") + scale_x_continuous(expand = expansion(mult = .2))
ven_bact 
```

```{r CoreMicrobiota}
core_microbiota <- intersect(get_taxa_unique(spain_understory_exclusive, taxonomic.rank = "Genus"), get_taxa_unique(senegal_understory_exclusive, taxonomic.rank = "Genus"))
pander(paste("Common genera between Senegal and Spain exclusive understory are: ", paste(core_microbiota, collapse=", ")))


core_microbiota_all <- intersect(get_taxa_unique(spain_understory_exclusive.all, taxonomic.rank = "Genus"), get_taxa_unique(senegal_understory_exclusive.all, taxonomic.rank = "Genus"))

pander(paste("Among the shared genera between Senegal and Spain exclusive understory, ", length(core_microbiota_all), "are within all samples: ", paste(core_microbiota_all, collapse=", ")))
```
